File: backend/app/routers/history.py
from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.services.history_service import delete_history, list_history, save_history, update_history

router = APIRouter(prefix="/api/history", tags=["history"])
TextMode = Literal["normal", "office", "study", "prompt"]


class HistoryCreateRequest(BaseModel):
    text: str = Field(..., min_length=1)
    mode: TextMode = "normal"


class HistoryUpdateRequest(BaseModel):
    text: str = Field(..., min_length=1)
    mode: TextMode | None = None


@router.get("")
def get_history() -> dict[str, object]:
    return {
        "code": 200,
        "message": "success",
        "data": list_history(),
    }


@router.post("")
def create_history(request: HistoryCreateRequest) -> dict[str, object]:
    try:
        record = save_history(request.text, request.mode)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error

    return {
        "code": 200,
        "message": "history saved",
        "data": record,
    }


@router.patch("/{record_id}")
def patch_history(record_id: int, request: HistoryUpdateRequest) -> dict[str, object]:
    try:
        record = update_history(record_id, request.text, request.mode)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error

    return {
        "code": 200,
        "message": "history updated",
        "data": record,
    }


@router.delete("/{record_id}")
def remove_history(record_id: int) -> dict[str, object]:
    try:
        delete_history(record_id)
    except ValueError as error:
        raise HTTPException(status_code=404, detail=str(error)) from error

    return {
        "code": 200,
        "message": "history deleted",
        "data": {"id": record_id},
    }

File: frontend/src/api/history.ts
const API_BASE_URL = "http://127.0.0.1:8000";

export type HistoryRecord = {
  id: number;
  text: string;
  mode: "normal" | "office" | "study";
  created_at: string;
};

type HistoryListResponse = {
  data: HistoryRecord[];
};

export async function fetchHistory() {
  const response = await fetch(`${API_BASE_URL}/api/history`);
  if (!response.ok) {
    throw new Error("获取历史记录失败。");
  }

  const result = (await response.json()) as HistoryListResponse;
  return result.data;
}

export async function saveHistory(text: string, mode: HistoryRecord["mode"]) {
  const response = await fetch(`${API_BASE_URL}/api/history`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text, mode }),
  });

  if (!response.ok) {
    throw new Error("保存历史记录失败。");
  }
}
